cbits/cryptonite_aes.c

```diff
@@ -406,6 +438,139 @@ void cryptonite_aes_gcm_finish(uint8_t *tag, aes_gcm *gcm, aes_key *key)
 	}
 }
 
+static inline int ccm_b0_flags(int has_adata, int m, int l)
+{
+	return 8*m + l + (has_adata? 64: 0);
+}
+
+/* depends on input size */
+static void ccm_encode_b0(block128* output, aes_ccm* ccm, int has_adata)
+{
+	int last = 15;
+	int m = ccm->length_M;
+	int l = ccm->length_L;
+	unsigned msg_len = ccm->length_input;
+
+	block128_zero(output);
+	block128_copy(output, &ccm->nonce);
+	output->b[0] = ccm_b0_flags(has_adata, (m-2)/2, l-1);
+	while (msg_len > 0) {
+		output->b[last--] = msg_len & 0xff;
+		msg_len >>= 8;
+	}
+}
+
+/* encode adata length */
+static int ccm_encode_la(block128* output, unsigned la)
+{
+	if (la < ( (1 << 16) - (1 << 8)) ) {
+		output->b[0] = (la >> 8) & 0xff;
+		output->b[1] = la        & 0xff;
+		return 2;
+	} else if (la < (1ull << 32)) {
+		output->b[0] = 0xff;
+		output->b[1] = 0xfe;
+		output->b[2] = (la >> 24) & 0xff;
+		output->b[3] = (la >> 16) & 0xff;
+		output->b[4] = (la >>  8) & 0xff;
+		output->b[5] = la         & 0xff;
+		return 6;
+	}
+}
+
+static void ccm_encode_ctr(block128* out, aes_ccm* ccm, unsigned int cnt)
+{
+	int last = 15;
+	block128_copy(out, &ccm->nonce);
+	out->b[0] = ccm->length_L - 1;
+
+	while (cnt > 0) {
+		out->b[last--] = cnt & 0xff;
+		cnt >>= 8;
+	}
+}
+
+static void ccm_cbcmac_add(aes_ccm* ccm, aes_key* key, block128* bi)
+{
+	block128_xor(&ccm->xi, bi);
+	cryptonite_aes_generic_encrypt_block(&ccm->xi, key, &ccm->xi);
+}
+
+/* even though it is possible to support message size as large as 2^64, we support up to 2^32 only */
+void cryptonite_aes_ccm_init(aes_ccm *ccm, aes_key *key, uint8_t *nonce, uint32_t nonce_len, uint32_t input_size, int m, int l)
+{
+	memset(ccm, 0, sizeof(aes_ccm));
+
+	if (l < 2 || l > 4) return;
+	if (m != 4 && m != 6 && m != 8 && m != 10
+		   && m != 12 && m != 14 && m != 16) return;
+
+	if (nonce_len != 15 - l) {
+		nonce_len = 15 - l;
+	}
```

Comment:
Not sure what to do with this.

We could ensure `nonce_len >= 15 - l` (ignoring extra bytes) and move the test to Haskell so `ccmInit` returns a failure when `nonce_len` is too short.

Function `cryptonite_aes_ocb_init` already has something like this:
```c
        /* we don't accept more than 15 bytes, any bytes higher will be ignored. */
        if (len > 15) {
                len = 15;
        }
```

Reply:
Should it be changed to if (nonce_len >= 15 - l) nonce_len = 15 -l instead? Because ccmInit returns a concrete type (instead of Maybe/Either), it isn't straight forward to handle errors unless throw exceptions. Do you have any suggestions so that errors can be handle more gratefully?

Reply:
Here's how I view things:
1. at C level priority is to ensure we don't read/write outside allocated buffers. So the `memcpy` call can be changed too, so that `nonce_len` is now used
2. in Haskell `ccmInit` and `aeadInit` can return CryptoError_IvSizeInvalid when the size is not compatible with L